### PR TITLE
lite client cmds

### PR DESCRIPTION
**Pull request by cmwaters**
in response to #7 and #1

Made some of the changes that Jack suggested including using `getChain` instead of using a liteClient map, converted cmds into function in `relayer/verifier.go`, removed most function level variables

### CODE REVIEW

File: cmd/lite.go [hunk 1]
```diff
@@ -17,11 +17,23 @@ package cmd
...
-	lite "github.com/tendermint/tendermint/lite2"
+)
+
+var (
+	//	lcMap map[string]*lite.Client // chainID => client
+	//
+	//	trustedHash    []byte
+	//	trustedHeight  int64
+	//	trustingPeriod time.Duration
+	updatePeriod time.Duration
```
Comment (jackzampolin) [inline cmd/lite.go]: This is nicer as a flag than passed in via config 👍 ✨

File: cmd/lite.go [hunk 2]
```diff
@@ -30,82 +42,135 @@ var liteCmd = &cobra.Command{
...
 	Short: "This command starts the auto updating relayer and logs when new headers are recieved",
 	Args:  cobra.ExactArgs(1),
 	RunE: func(cmd *cobra.Command, args []string) error {
 		chainID := args[0]
-
-		if !relayer.Exists(chainID, config.c) {
-			return fmt.Errorf("chain with ID %s is not configured", chainID)
-		}
-
+		updatePeriod, _ = cmd.Flags().GetDuration("update-period")
```
Comment (jackzampolin) [inline cmd/lite.go]: Would be nice to have that `update-period` string in a package level var.

File: cmd/lite.go [hunk 3]
```diff
@@ -30,82 +42,135 @@ var liteCmd = &cobra.Command{
...
-				// reobtain trust height and hash
-				return err
-			default:
-				// try with another full node
-				return err
-			}
-		}
-
-		return nil
+		return chain.StartLiteClient(filepath.Join(liteDir, chainID), updatePeriod)
```
Comment (jackzampolin) [inline cmd/lite.go]: Does this call block? Is there any logging here?

File: cmd/lite.go [hunk 4]
```diff
@@ -30,82 +42,135 @@ var liteCmd = &cobra.Command{
...
 }
 
-// TODO: Figure out arguements for initializing a lite client with a root of trust
-var liteUpdateCmd = &cobra.Command{
-	Use:   "update [chain-id] [hash] [height] [trust-period]",
-	Short: "Update an existing lite client for a configured chain, requres passing in a root of trust",
-	Args:  cobra.ExactArgs(3),
+// Only checks headers in the trusted store - doesn't validate headers from the trusted node that don't yet exist in the lite client itself
+var liteGetHeaderCmd = &cobra.Command{
+	Use: "header [chain-id] [height]",
```
Comment (jackzampolin) [inline cmd/lite.go]: Would be nice to have a command to fetch roots of trust (headers) and valsets from an endpoint too.

File: relayer/verifier.go [hunk 5]
```diff
@@ -2,89 +2,47 @@ package relayer
...
-	"github.com/tendermint/tendermint/lite2/provider/http"
 )
 
-// StartAutoLiteClient creates and starts an auto lite client
-func (c *Chain) StartAutoLiteClient(homePath, homeDir string, cache int) error {
-	autoLite, err := c.NewAutoLiteClient(homePath, homeDir, cache)
-	if err != nil {
-		return err
+// Spins up an instance of the lite client as part of the chain.
+func (c *Chain) StartLiteClient(dbDir string, updatePeriod time.Duration) error {
```
Comment (jackzampolin) [inline relayer/verifier.go]: How to run this in a goroutine?

**Comment by melekes** [comment]
https://github.com/cosmos/relayer/pull/3#pullrequestreview-347506317 review should be taken into account

File: cmd/lite.go (hunk 3 above)
Comment (melekes) [reply cmd/lite.go]: `start` (needs to be renamed to `reset`) cmd should call an HTTP endpoint I believe https://github.com/cosmos/relayer/issues/5#issuecomment-578656671

File: cmd/lite.go [hunk 6]
```diff
@@ -30,82 +42,135 @@ var liteCmd = &cobra.Command{
...
-	Use:   "init [chain-id] [hash] [height] [trust-period]",
-	Short: "Create a new lite client for a configured chain, requires passing in a root of trust",
+var liteDeleteCmd = &cobra.Command{
+	Use:   "delete [chain-id]",
+	Short: "Delete an existing lite client for a configured chain, this will force new initialization during the next usage of the lite client.",
 	Args:  cobra.ExactArgs(1),
 	RunE: func(cmd *cobra.Command, args []string) error {
-		return nil
+		chainID := args[0]
+		chain, err := relayer.GetChain(chainID, config.c)
```
Comment (melekes) [inline cmd/lite.go]: same here, but for now, we can leave these to empty

File: cmd/lite.go [hunk 7]
```diff
@@ -30,82 +42,135 @@ var liteCmd = &cobra.Command{
...
 	},
 }
 
-// TODO: Figure out arguements for initializing a lite client with a root of trust
-var liteUpdateCmd = &cobra.Command{
-	Use:   "update [chain-id] [hash] [height] [trust-period]",
-	Short: "Update an existing lite client for a configured chain, requres passing in a root of trust",
-	Args:  cobra.ExactArgs(3),
+// Only checks headers in the trusted store - doesn't validate headers from the trusted node that don't yet exist in the lite client itself
+var liteGetHeaderCmd = &cobra.Command{
```
Comment (melekes) [inline cmd/lite.go]:
```suggestion
var liteHeaderCmd = &cobra.Command{
```

File: cmd/lite.go [hunk 8]
```diff
@@ -30,82 +42,135 @@ var liteCmd = &cobra.Command{
...
 	RunE: func(cmd *cobra.Command, args []string) error {
+		chainID := args[0]
+		chain, err := relayer.GetChain(chainID, config.c)
+		if err != nil {
+			return err
+		}
+		if chain.LiteClient != nil {
+			return fmt.Errorf("lite client is not running on this chain")
+		}
+		if len(args) == 1 {
```
Comment (melekes) [inline cmd/lite.go]:
```go
switch len(args) {
  case 1:
  case 2:
  default:
    return errors.New("expected either 1 or 2 arguments")
}
```

File: cmd/lite.go [hunk 9]
```diff
@@ -30,82 +42,135 @@ var liteCmd = &cobra.Command{
...
+		chainID := args[0]
+		chain, err := relayer.GetChain(chainID, config.c)
+		if err != nil {
+			return err
+		}
+		if chain.LiteClient != nil {
+			return fmt.Errorf("lite client is not running on this chain")
+		}
+		lc := chain.LiteClient
+		if len(args) == 1 { // return latest validator set
```
Comment (melekes) [inline cmd/lite.go]:
```go
switch len(args) {
  case 1:
  case 2:
  default:
    return errors.New("expected either 1 or 2 arguments")
}
```

File: cmd/lite.go [hunk 10]
```diff
@@ -30,82 +42,135 @@ var liteCmd = &cobra.Command{
...
+		}
 		return nil
 	},
 }
 
-var liteDeleteCmd = &cobra.Command{
-	Use:   "delete [chain-id]",
-	Short: "Delete an existing lite client for a configured chain, this will force new initialization during the next usage of the lite clien.",
-	Args:  cobra.ExactArgs(1),
+var liteGetValidatorsCmd = &cobra.Command{
```
Comment (melekes) [inline cmd/lite.go]:
```suggestion
var liteValidatorsCmd = &cobra.Command{
```

https://golang.org/doc/effective_go.html#Getters

File: cmd/lite.go [hunk 11]
```diff
@@ -30,82 +42,135 @@ var liteCmd = &cobra.Command{
...
+		chainID := args[0]
+		chain, err := relayer.GetChain(chainID, config.c)
+		if err != nil {
+			return err
+		}
+		if chain.LiteClient != nil {
+			return fmt.Errorf("lite client is not running on this chain")
+		}
+		if len(args) == 1 {
+			header, err := chain.LiteClient.TrustedHeader(0, time.Now())
```
Comment (melekes) [inline cmd/lite.go]: You need to use the DB directly. Initializing a new client just to query something is not an option. You can initialize the DB store here https://godoc.org/github.com/tendermint/tendermint/lite2/store/db and use it to query headers / validator sets

File: cmd/lite.go [hunk 12]
```diff
@@ -30,82 +42,135 @@ var liteCmd = &cobra.Command{
...
 	RunE: func(cmd *cobra.Command, args []string) error {
+		chainID := args[0]
+		chain, err := relayer.GetChain(chainID, config.c)
+		if err != nil {
+			return err
+		}
+		if chain.LiteClient != nil {
+			return fmt.Errorf("lite client is not running on this chain")
+		}
+		lc := chain.LiteClient
```
Comment (melekes) [inline cmd/lite.go]: You need to use the DB directly. Initializing a new client just to query something is not an option. You can initialize the DB store here https://godoc.org/github.com/tendermint/tendermint/lite2/store/db and use it to query headers / validator sets

File: cmd/lite.go (hunk 2 above)
Comment (cmwaters) [reply cmd/lite.go]: update-period is now part of the config file

File: cmd/lite.go (hunk 8 above)
Comment (cmwaters) [reply cmd/lite.go]: What's the rationale behind using this?

Comment (melekes) [reply cmd/lite.go]: switch has a better structure

File: cmd/config.go [hunk 13]
```diff
@@ -74,7 +75,12 @@ func setChains(c *Config) error {
 		for _, cp := range i.Counterparties {
 			cps = append(cps, relayer.NewCounterparty(cp.ChainID, cp.ClientID))
 		}
-		chain, err := relayer.NewChain(i.Key, i.ChainID, i.RPCAddr, i.AccountPrefix, cps, i.Gas, i.GasAdjustment, i.GasPrices, i.DefaultDenom, i.Memo, homePath, c.Global.LiteCacheSize, i.TrustOptions)
+		updatePeriod, err := time.ParseDuration(i.UpdatePeriod)
```
Comment (jackzampolin) [inline cmd/config.go:80]: Maybe we do this check as we read in the config and set it on the `Chain` abstraction?

File: cmd/lite.go (hunk 11 above)
Comment (jackzampolin) [reply cmd/lite.go]: Seems like this functionality should be in the `Chain` abstraction

File: cmd/start.go [hunk 14]
```diff
@@ -33,9 +34,16 @@ var startCmd = &cobra.Command{
 			return err
 		}
 
-		// The relayer will
-		for {
+		// initiate a lite client for each chain declared in the config file
```
Comment (jackzampolin) [inline cmd/start.go]: Nice!

File: relayer/chain.go [hunk 15]
```diff
@@ -79,10 +82,12 @@ type Chain struct {
...
-	Keybase    keys.Keybase
-	Client     *rpcclient.HTTP
-	Cdc        *codec.Codec
-	LiteClient *lite.Client
+	Keybase      keys.Keybase
+	Client       *rpcclient.HTTP
+	Cdc          *codec.Codec
+	LiteClient   *lite.Client
+	UpdatePeriod time.Duration
+	logger       log.Logger
```
Comment (jackzampolin) [inline relayer/chain.go]: Are we setting/using this logger anywhere? Seems like this could be used to give chain specific error messages.

Comment (cmwaters) [reply relayer/chain.go]: Yeah I started with some of the chain abstraction:
```
func (c *Chain) LatestHeight() uint64 {
	height, err := c.LiteClient.LastTrustedHeight()
	if err != nil {
		c.logger.Error("lite client does not have a last trusted height", err)
		height = 0
	}
	return uint64(height)
}
```

Comment (jackzampolin) [reply relayer/chain.go]: Nice! I will integrate this more deeply. Doesn't need to be in scope for this PR.

File: cmd/config.go [hunk 16]
```diff
@@ -74,7 +75,12 @@ func setChains(c *Config) error {
...
+		if err != nil {
+			return nil
```
Comment (fedekunze) [inline cmd/config.go:82]:
```suggestion
			return err
```

File: cmd/config.go [hunk 17]
```diff
@@ -74,7 +75,12 @@ func setChains(c *Config) error {
...
+		}
+		chain, err := relayer.NewChain(i.Key, i.ChainID, i.RPCAddr, i.AccountPrefix, cps, i.Gas, i.GasAdjustment,
+			i.GasPrices, i.DefaultDenom, i.Memo, homePath, c.Global.LiteCacheSize, i.TrustOptions, updatePeriod)
 		if err != nil {
 			return nil
```
Comment (fedekunze) [inline cmd/config.go:88]:
```suggestion
			return err
```

File: relayer/verifier.go [hunk 18]
```diff
@@ -2,89 +2,44 @@ package relayer
...
 
-// StartAutoLiteClient creates and starts an auto lite client
-func (c *Chain) StartAutoLiteClient(homePath, homeDir string, cache int) error {
-	autoLite, err := c.NewAutoLiteClient(homePath, homeDir, cache)
-	if err != nil {
-		return err
+// Spins up an instance of the lite client as part of the chain.
+func (c *Chain) StartLiteClient(dbDir string) error {
+	if c.LiteClient != nil {
+		return fmt.Errorf("instance of lite client already running for this chain")
```
Comment (fedekunze) [inline relayer/verifier.go]:
```suggestion
		return errors.New("instance of lite client already running for this chain")
```

File: relayer/verifier.go [hunk 19]
```diff
@@ -2,89 +2,44 @@ package relayer
...
-	db, err := dbm.NewGoLevelDB(path, homeDir)
-	if err != nil {
-		return out, err
-	}
+	c.LiteClient, err = lite.NewClient(c.ChainID,
+		c.TrustOptions.Get(),
+		httpProvider,
+		dbs.New(db, c.ChainID),
+		lite.UpdatePeriod(c.UpdatePeriod))
+	return err
```
Comment (fedekunze) [inline relayer/verifier.go]:
```suggestion
	if err != nil {
	    return err
	}
	return nil
```

**Review by fedekunze — CHANGES_REQUESTED** [review]
👍

**Comment by jackzampolin** [comment]
Pushed some changes here. Going to go ahead and merge it. Let's continue the work on another branch. While working with the code one thing I noticed was that there is some confusion over what the various lite client commands should be doing. Lets grab some time next week to clear that up.

**Pull request merged**